### ISSUE TITLE
Simplied Format for Communicating Connectivity

### ISSUE THREAD

**Issue by snasphysicist** [biomakespace/OpenPlantToy#12]
Currently the connectivity of the circuit is communicated to the controller by the units assembling a long string containing all connectivity information passed along the whole tree. This is a bit slow and inefficient, there is a risk of overloading the 64 byte serial buffers with a single transfer and it complicates the code & potentially introduces timing issues since units are having to wait for communication from downstream to assemble the string in a consistent manner.

A potentially better way: suppose we have connection CONTROLLER - ID1 - ID2 - ID3
- ID3 passes string ID3; to ID2
- ID2 strips the id ID3, throws away the ; and passes downstream anything left in the string (here, nothing)
- ID2 creates the string ID2;ID2-ID3, to indicate it is attached to ID3 with ID3 upstream
- ID2 passes this string to ID1
- ID1 strips out ID2, throws away the ;, and passes the rest (ID2-ID3,) downstream
- ID1 constructs string ID1;ID1-ID2, to indicate is is attached to ID2 upstream
- ID1 passes this string downstream

Basically:
- If the unit receives a string from upstream with no semicolon, just pass it along
- If the unit receives a string from upstream with a semicolon, remove/store the preceding ID & ;, pass along the rest, then construct a new string with the unit's ID and the ID it just stored and send this downstream

Now the connectivity of the units is communicated to the controller one pair at a time, and the burden is on the controller to figure out how to put this back together (as it should be, since it is currently the most computationally powerful part of the setup).

**Comment by snasphysicist** [comment]
This won't work well, because it doesn't avoid the need for tracking whether or not the unit is connected to something upstream.

What will work better - as soon as a unit (ID1) receives a message from downstream, it should pass this along and then send a message downstream containing its id like ID1;

When a unit (ID2) receives a message from upstream like ID1; it should construct the string ID2-ID1, and pass this downstream.

When a unit receives a message from upstream like ID2-ID1, it will just pass it downstream.

Because each unit independently sends its ID downstream, it is not necessary for message passing downstream to be 'initiated' by the furthermost unit (it is a pain to work out which unit is furthermost in this scheme).

Then the controller should just junk any message containing a semicolon (it *should* only receive one) and assemble the connections from the messages containing commas.

**Comment by snasphysicist** [comment]
I have tested this with two components and it's working quite nicely, seems much smoother. Test with more components required. Merged into develop in 6703d63 but not closed pending further testing.

**Comment by snasphysicist** [comment]
Also: documentation updates will be required
